zjit/src/codegen.rs

```diff
@@ -890,14 +890,23 @@ fn gen_new_range(
 }
 
 /// Compile code that exits from JIT code with a return value
-fn gen_return(jit: &JITState, asm: &mut Assembler, val: lir::Opnd) -> Option<()> {
+fn gen_return(jit: &JITState, asm: &mut Assembler, mut val: lir::Opnd) -> Option<()> {
     // Pop the current frame (ec->cfp++)
     // Note: the return PC is already in the previous CFP
     asm_comment!(asm, "pop stack frame");
     let incr_cfp = asm.add(CFP, RUBY_SIZEOF_CONTROL_FRAME.into());
     asm.mov(CFP, incr_cfp);
     asm.mov(Opnd::mem(64, EC, RUBY_OFFSET_EC_CFP), CFP);
 
+    // Load any stack relative memory since we're about to destroy the current frame
+    const NATIVE_SP_REG_NO: u8 = {
+        let Opnd::Reg(reg) = NATIVE_STACK_PTR else { unreachable!(); };
+        reg.reg_no
+    };
+    if let Opnd::Mem(lir::Mem { base: lir::MemBase::Reg(NATIVE_SP_REG_NO), .. }) = val {
```

Comment:
Not sure I am following. Don't we also need to check this post-regalloc then?

Reply:
We'd need to check after regalloc if it supported spilling, but it doesn't right now. The SP-relative operands here are coming from param_opnd() in the same file.

https://github.com/ruby/ruby/blob/6b0e5de4e68bc60c10ce7974cb69f246678a9512/zjit/src/codegen.rs#L1110-L1119

We refer to the same `lir::Opnd` for each param.

The does feel like something LIR should provide an API for so we can't get it wrong. I'm adding a `MemBase::FrameBase` so we can refer to the right slot even after operations like cpush, but that won't get us all the way there.

Reply:
We support at least block spilling now, though, after @k0kubun 's change

Reply:
Yeah. It should be good for now.